src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -4592,6 +4592,41 @@ public override BoundNode VisitThrowExpression(BoundThrowExpression node)
             return result;
         }
 
+        public override BoundNode VisitYieldReturnStatement(BoundYieldReturnStatement node)
+        {
+            BoundExpression expr = node.Expression;
+            if (expr == null)
+            {
+                return null;
+            }
+            var method = (MethodSymbol)_member;
+            Debug.Assert(method.IsIterator);
+            VisitOptionalImplicitConversion(expr, getIteratorElementType(method.ReturnType.TypeSymbol), useLegacyWarnings: false, AssignmentKind.Return);
+            return null;
+
+            TypeSymbolWithAnnotations getIteratorElementType(TypeSymbol returnType)
+            {
+                switch (returnType.OriginalDefinition.SpecialType)
+                {
+                    case SpecialType.System_Collections_IEnumerable:
+                    case SpecialType.System_Collections_IEnumerator:
+                        return TypeSymbolWithAnnotations.Create(compilation.GetSpecialType(SpecialType.System_Object));
+
+                    case SpecialType.System_Collections_Generic_IEnumerable_T:
+                    case SpecialType.System_Collections_Generic_IEnumerator_T:
+                        return ((NamedTypeSymbol)returnType).TypeArgumentsNoUseSiteDiagnostics[0];
+                }
+
+                if (returnType.OriginalDefinition == compilation.GetWellKnownType(WellKnownType.System_Collections_Generic_IAsyncEnumerable_T))
+                {
+                    return ((NamedTypeSymbol)returnType).TypeArgumentsNoUseSiteDiagnostics[0];
+                }
+
+                Debug.Assert(false);
```

Comment:
Consider throwing `ExceptionUtilities.Unreachable`?

Reply:
Had to adjust this further for regressions caught by CI. Now `return default;`. Sorry for the churn.
The scenario is that you put a `yield return` in a regular method.

---
In reply to: [232775143](https://github.com/dotnet/roslyn/pull/31072#discussion_r232775143) [](ancestors = 232775143)